### PR TITLE
[JUJU-4504] Retry transactions when entering scope of a relation unit

### PR DESCRIPTION
**Pull request by nvinuesa**
_TLDR: Under some load (a few apps and units) sometimes a newly created relation fails to enter scope with error `inconsistent state in EnterScope`. This is due to the transactions involved being aborted with `WriteConflict error`. 
The main hypothesis behind this patch is that this error is purely transient (due to transactions that write to the same document being run concurrently, and on different transaction runners) and can therefore be retried._

This patch fixes a bug (https://bugs.launchpad.net/juju/+bug/2031631) on which the relation unit doesn't enter scope in the case multiple units and applications are present before creating a relation on them.

When using server-side transactions, mgo retries (hard-coded 3 times) transactions that fail with WriteConflict error, and after that it aborts the (mgo) transaction.
For example:
```
juju debug-log --replay --level TRACE --tail -m controller | grep "juju.txn.*Please retry"
controller-0: 15:44:17 TRACE juju.txn update op: txn.Op{C:"modelEntityRefs", Id:"03b60bbf-53e8-48e9-819c-dfaf1e71b954", Assert:"d+", Insert:interface {}(nil), Update:bson.D{bson.DocElem{Name:"$addToSet", Value:bson.D{bson.DocElem{Name:"applications", Value:"database"}}}}, Remove:false} write conflict: WriteConflict error: this operation conflicted with another operation. Please retry your operation or multi-document transaction.
```
These retries are seen many times on multiple scenarios in juju, and therefore the business logic decides on each case how to treat the (mgo server-side) aborted transactions.

In the case of EnterScope, it is necessary to retry the (mgo server-side) transaction (which includes retrying the mongodb transaction) because EnterScope is called concurrently (via the EnterScope facade) and therefore WriteConflict errors are transient errors. By calling Run(jujutxn.TransactionSource) instead of RunTransaction(ops []txn.Op) we make use of the retries applied to the function (jujutxn.TransactionSource) passed to Run(...).

Note: This last observation means that the true solution would be sequential transactions for EnterScope, but since each relationunit has it's own runner, this would mean a huge change.

### Reproducing the bug

The first way to reproduce this behaviour is simply by modifying the `func (prr *ProReqRelation) allEnterScope(c *gc.C)` function on `state/relationunit_test.go` in a way that the calls to `EnterScope()` are run concurrently:
```go
func (prr *ProReqRelation) allEnterScope(c *gc.C) {
       g, _ := errgroup.WithContext(context.Background())

       g.Go(func() error { return prr.pru0.EnterScope(nil) })
       g.Go(func() error { return prr.pru1.EnterScope(nil) })
       g.Go(func() error { return prr.rru0.EnterScope(nil) })
       g.Go(func() error { return prr.rru1.EnterScope(nil) })

       err := g.Wait()
	c.Assert(err, jc.ErrorIsNil)
}
```
This change only should make some tests fail, for example:
```
go test github.com/juju/juju/state/... -gocheck.v -gocheck.f=TestWatchAppSettings
...
[LOG] 0:01.324 DEBUG juju.state.pool.txnwatcher txn watcher: relations eea4884f-5783-4a66-897a-5d55f6a41daf:wordpress:db mysql:server #5
    prr.allEnterScope(c)
relationunit_test.go:1081:
    c.Assert(err, jc.ErrorIsNil)
... value *errors.errorString = &errors.errorString{s:"cannot enter scope for unit \"wordpress/1\" in relation \"wordpress:db mysql:server\": inconsistent state in EnterScope"} ("cannot enter scope for unit \"wordpress/1\" in relation \"wordpress:db mysql:server\": inconsistent state in EnterScope")
...
```
Another way is by bootstrapping a microk8s controller, adding a model and simply deploying two apps with multiple units and a relation between them, it won't fail systematically but if you delete and recreate the relation it will eventually fail:
```
juju deploy kafka-k8s -n 5 --channel 3/edge && juju deploy zookeeper-k8s -n 5 --channel 3/edge && juju relate kafka-k8s zookeeper-k8s
juju debug-log --replay --level DEBUG --tail | grep "enterscope\|EnterScope" 
```
The last way is by reproducing the scenario in https://bugs.launchpad.net/juju/+bug/2031631. This is the longest setup but it failed systematically locally on my PC:

```
# using juju 3.1 (from snap) bootstrap a microk8s controller and add a model
juju bootstrap microk8s c --config logging-config="<root>=DEBUG;juju.txn=TRACE;juju.worker.uniter.relation=TRACE"
juju add-model testing
# clone the data project and checkout the correct PR:
git clone git@github.com:canonical/data-platform-libs.git && gh pr checkout 84 
# setup python virtual environment
python3 -m venv data-platform-libs
# activate the virtual env and install dependencies
cd data-platform-libs
source bin/activate 
pip install -r requirements.txt
# set the correct pylibjuju version and run the tests
export LIBJUJU_VERSION_SPECIFIER="==3.2.0.1"
tox run -e integration-database -- -m 'not unstable' --model testing
```
by checking the logs you should see the error:
```
juju debug-log --replay --level DEBUG --tail | grep "enterscope\|EnterScope"
unit-database-1: 15:45:33 ERROR juju.worker.uniter resolver loop error: cannot enter scope for unit "database/1" in relation "application:first-database database:database": inconsistent state in EnterScope
unit-database-1: 15:45:33 INFO juju.worker.uniter unit "database/1" shutting down: cannot enter scope for unit "database/1" in relation "application:first-database database:database": inconsistent state in EnterScope
unit-database-1: 15:45:33 DEBUG juju.worker.dependency "uniter" manifold worker stopped: cannot enter scope for unit "database/1" in relation "application:first-database database:database": inconsistent state in EnterScope
cannot enter scope for unit "database/1" in relation "application:first-database database:database": inconsistent state in EnterScope
github.com/juju/juju/api/agent/uniter.(*RelationUnit).EnterScope:71:
```

## Checklist

*If an item is not applicable, use `~strikethrough~`.*

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

QA steps include to run the scenarios described in the `Reproducing the bug` section above. These should always pass now.

## Bug reference

https://bugs.launchpad.net/juju/+bug/2031631

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -303,3 +303,5 @@ replace gopkg.in/yaml.v2 => github.com/juju/yaml/v2 v2.0.0
 replace github.com/dustin/go-humanize v1.0.0 => github.com/dustin/go-humanize v0.0.0-20141228071148-145fabdb1ab7
 
 replace github.com/hashicorp/raft-boltdb => github.com/juju/raft-boltdb v0.0.0-20200518034108-40b112c917c5
+
+replace github.com/juju/txn => /home/nicolas/workspace/canonical/txn
```
Comment (jameinel) [inline go.mod:307]: We probably don't want to commit this one.

File: state/relationunit.go [hunk 2]
```diff
@@ -76,133 +76,138 @@ func (ru *RelationUnit) EnterScope(settings map[string]interface{}) error {
...
-			return stateerrors.ErrCannotEnterScope
+			// The relation or unit might no longer be Alive. (Note that there is no
+			// need for additional checks if we're trying to create a subordinate
+			// unit: this could fail due to the subordinate applications not being Alive,
+			// but this case will always be caught by the check for the relation's
+			// life (because a relation cannot be Alive if its applications are not).)
+			relations, rCloser := db.GetCollection(relationsC)
+			defer rCloser()
+			if alive, err := isAliveWithSession(relations, relationDocID); err != nil {
+				return nil, err
```
Comment (jameinel) [inline state/relationunit.go:105]: we should probably be errors.Trace(err) for all of these.

File: state/relationunit.go [hunk 3]
```diff
@@ -76,133 +76,138 @@ func (ru *RelationUnit) EnterScope(settings map[string]interface{}) error {
...
+			} else if !alive {
+				return nil, stateerrors.ErrCannotEnterScope
```
Comment (jameinel) [inline state/relationunit.go:107]: Do we want something like:
errors.Annotate(stateerrors.ErrCannotEnterScope, "relation is no longer alive")

File: state/relationunit.go [hunk 4]
```diff
@@ -76,133 +76,138 @@ func (ru *RelationUnit) EnterScope(settings map[string]interface{}) error {
...
+			}
+			if ru.isLocalUnit {
+				units, uCloser := db.GetCollection(unitsC)
+				defer uCloser()
+				if alive, err := isAliveWithSession(units, ru.unitName); err != nil {
+					return nil, err
+				} else if !alive {
+					return nil, stateerrors.ErrCannotEnterScope
```
Comment (jameinel) [inline state/relationunit.go:115]: errors.Trace
and
errors.Annotatef(ErrCannotEnterScope, "unit %v is no longer alive", ru.unitName)

File: state/relationunit.go [hunk 5]
```diff
@@ -76,133 +76,138 @@ func (ru *RelationUnit) EnterScope(settings map[string]interface{}) error {
...
+				}
+
+				// Maybe a subordinate used to exist, but is no longer alive. If that is
+				// case, we will be unable to enter scope until that unit is gone.
+				if existingSubName != "" {
+					if alive, err := isAliveWithSession(units, existingSubName); err != nil {
+						return nil, err
+					} else if !alive {
+						return nil, stateerrors.ErrCannotEnterScopeYet
```
Comment (jameinel) [inline state/relationunit.go:124]: why is this one "ErrCannotEnterScopeYet" vs "ErrCannotEnterScope" for the other ones?

We could also use an annotation to indicate what 'existingSubName' is not alive.

File: state/relationunit.go [hunk 6]
```diff
@@ -76,133 +76,138 @@ func (ru *RelationUnit) EnterScope(settings map[string]interface{}) error {
...
+					}
+				}
+			}
+
+			prefix := fmt.Sprintf("cannot enter scope for unit %q in relation %q: ", ru.unitName, ru.relation)
```
Comment (jameinel) [inline state/relationunit.go:129]: or maybe, you already have this prefix, should we put something like this as the annotation around all the other checks?

File: state/relationunit.go [hunk 7]
```diff
@@ -76,133 +76,138 @@ func (ru *RelationUnit) EnterScope(settings map[string]interface{}) error {
...
-		// case, we will be unable to enter scope until that unit is gone.
-		if existingSubName != "" {
-			if alive, err := isAliveWithSession(units, existingSubName); err != nil {
-				return err
-			} else if !alive {
-				return stateerrors.ErrCannotEnterScopeYet
+		// Verify that the unit is not already in scope, and abort without error
+		// if it is.
+		if count, err := relationScopes.FindId(ruKey).Count(); err != nil {
+			return nil, err
```
Comment (jameinel) [inline state/relationunit.go:144]: errors.Trace()

File: state/relationunit.go [hunk 8]
```diff
@@ -76,78 +76,84 @@ func (ru *RelationUnit) EnterScope(settings map[string]interface{}) error {
...
-	var ops []txn.Op
-	if ru.isLocalUnit {
+
+	settingsChanged := func() (bool, error) { return false, nil }
+
+	var existingSubName string
+
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+
+		// Verify that the unit is not already in scope, and abort without error
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply state/relationunit.go]: done

File: state/relationunit.go [hunk 9]
```diff
@@ -76,78 +76,84 @@ func (ru *RelationUnit) EnterScope(settings map[string]interface{}) error {
...
-	// Collect the operations necessary to enter scope, as follows:
-	// * Check unit and relation state, and incref the relation.
-	// * TODO(fwereade): check unit status == params.StatusActive (this
-	//   breaks a bunch of tests in a boring but noisy-to-fix way, and is
-	//   being saved for a followup).
 	relationDocID := ru.relation.doc.DocID
-	var ops []txn.Op
-	if ru.isLocalUnit {
+
+	settingsChanged := func() (bool, error) { return false, nil }
```
[the remark this answers is not in the document]
Comment (nvinuesa) [reply state/relationunit.go]: done

File: state/relationunit.go (hunk 3 above)
Comment (nvinuesa) [reply state/relationunit.go]: I changed the enter scope errors a little bit, adding annotations on most of them

File: state/relationunit.go (hunk 4 above)
Comment (nvinuesa) [reply state/relationunit.go]: same as abobe

File: state/relationunit.go (hunk 5 above)
Comment (nvinuesa) [reply state/relationunit.go]: so, I don't know why the `Yet`, it does look from the comments that it's a transient error until the subordinate get's removed (if I understood correctly) but I don't know if we are dealing with this error on the other side as a true transient error

File: state/relationunit.go (hunk 6 above)
Comment (nvinuesa) [reply state/relationunit.go]: I moved this prefix up and used it on annotated errors

File: state/relationunit.go (hunk 7 above)
Comment (nvinuesa) [reply state/relationunit.go]: done

**Comment by hpidcock** [comment]
```
gostress ./state.test -check.f=WatchRelationUnitsSuite.TestWatchAppSettings
5s: 0 runs so far, 0 failures
10s: 24 runs so far, 0 failures
15s: 25 runs so far, 0 failures
20s: 48 runs so far, 0 failures
25s: 72 runs so far, 0 failures
30s: 72 runs so far, 0 failures
35s: 96 runs so far, 0 failures
40s: 120 runs so far, 0 failures
45s: 120 runs so far, 0 failures
50s: 144 runs so far, 0 failures
```
```
gostress ./state.test -check.f=RelationSuite.TestDestroy
5s: 0 runs so far, 0 failures
10s: 0 runs so far, 0 failures
15s: 24 runs so far, 0 failures
20s: 24 runs so far, 0 failures
25s: 48 runs so far, 0 failures
30s: 48 runs so far, 0 failures
35s: 72 runs so far, 0 failures
40s: 72 runs so far, 0 failures
45s: 94 runs so far, 0 failures
50s: 96 runs so far, 0 failures
55s: 97 runs so far, 0 failures
```
```
gostress ./state.test -check.f="CleanupSuite.(TestCleanupForceDestroyedMachineWithContainer|TestForceDestroyUnitDestroysSubordinates|TestForceDestroyUnitLeavesRelations|TestForceDestroyRelationIncorrectUnitCount)"
5s: 0 runs so far, 0 failures
10s: 0 runs so far, 0 failures
15s: 24 runs so far, 0 failures
20s: 24 runs so far, 0 failures
25s: 48 runs so far, 0 failures
30s: 49 runs so far, 0 failures
35s: 72 runs so far, 0 failures
40s: 75 runs so far, 0 failures
45s: 96 runs so far, 0 failures
50s: 102 runs so far, 0 failures
55s: 120 runs so far, 0 failures
```

File: state/relationunit.go [hunk 10]
```diff
@@ -76,133 +76,138 @@ func (ru *RelationUnit) EnterScope(settings map[string]interface{}) error {
...
-	} else if count != 0 {
-		// The scope document exists, so we're actually already in scope.
-		return nil
-	}
+			// The relation or unit might no longer be Alive. (Note that there is no
+			// need for additional checks if we're trying to create a subordinate
+			// unit: this could fail due to the subordinate applications not being Alive,
+			// but this case will always be caught by the check for the relation's
+			// life (because a relation cannot be Alive if its applications are not).)
+			relations, rCloser := db.GetCollection(relationsC)
```
Comment (hpidcock) [inline state/relationunit.go:102]: Might make sense to move the `.GetCollections(...)` calls to outside the buildTxn func. Not gonna block it over this, since I've seen this in other places.

**Review by hpidcock — APPROVED** [review]
Happy with this here. We can backport to 2.9 in a seperate PR or do it here.

**Comment by nvinuesa** [comment]
Closed in favor of https://github.com/juju/juju/pull/16257
